File: BusquedaBinaria.pas
program BusquedaBinaria;

uses crt;

const
  N = 10;

type
  TArreglo = array[1..N] of integer;

var
  arr: TArreglo;
  objetivo: integer;
  inicio, fin, medio, pos, i: integer;

begin
  clrscr;

  { Inicializar arreglo ORDENADO }
  arr[1] := 2;   arr[2] := 5;   arr[3] := 8;   arr[4] := 12;  arr[5] := 16;
  arr[6] := 23;  arr[7] := 38;  arr[8] := 45;  arr[9] := 56;  arr[10] := 72;

  writeln('Arreglo ordenado:');
  for i := 1 to N do
    write(arr[i]:4);
  writeln;

  { Leer número a buscar }
  write('Ingrese el numero a buscar: ');
  readln(objetivo);

  { Inicializar variables }
  inicio := 1;
  fin := N;
  pos := -1;

  { Algoritmo de búsqueda binaria }
  while (inicio <= fin) and (pos = -1) do
  begin
    medio := (inicio + fin) div 2;
    if arr[medio] = objetivo then
      pos := medio
    else if arr[medio] < objetivo then
      inicio := medio + 1
    else
      fin := medio - 1;
  end;

  { Mostrar resultado }
  if pos = -1 then
    writeln('Numero no encontrado en el arreglo.')
  else
    writeln('Numero encontrado en la posicion: ', pos);

  readln;
end.

File: Busqueda_Binaria.f
            PROGRAM BUSQUEDA_BINARIA
            
      PARAMETER (N = 10)
      INTEGER ARR(N), OBJETIVO
      INTEGER INICIO, FIN, MEDIO, POS, I

      DATA ARR /2, 5, 8, 12, 16, 23, 38, 45, 56, 72/

      PRINT *, 'Arreglo ordenado:'
      PRINT *, (ARR(I), I = 1, N)

      PRINT *, 'Ingrese el numero a buscar:'
      READ *, OBJETIVO

      INICIO = 1
      FIN = N
      POS = -1

 10   IF (INICIO .LE. FIN) THEN
       MEDIO = (INICIO + FIN) / 2
       IF (ARR(MEDIO) .EQ. OBJETIVO) THEN
         POS = MEDIO
         GOTO 20
       ELSE IF (ARR(MEDIO) .LT. OBJETIVO) THEN
         INICIO = MEDIO + 1
       ELSE
         FIN = MEDIO - 1
       END IF
       GOTO 10
      END IF

 20   IF (POS .EQ. -1) THEN
       PRINT *, 'Numero no encontrado en el arreglo.'
      ELSE
       PRINT *, 'Numero encontrado en la posicion:', POS
      END IF
      
      PAUSE

      END

